galley/pkg/config/source/kube/apiserver/source.go

```diff
@@ -79,6 +99,106 @@ func (s *Source) Start() {
 	}
 	s.started = true
 
+	// Create a set of pending resources. These will be matched up with incoming CRD events for creating watchers for
+	// each resource that we expect.
+	s.expectedResources = make(map[string]schema.KubeResource)
+	for _, r := range s.options.Resources {
+		// If we received the metadata with the resource marked as disabled, simply ignore it.
+		if r.Disabled {
+			continue
+		}
+
+		// Use the disabled bit to track whether we received an event for this resource.
+		r.Disabled = true
+		s.expectedResources[asKey(r.Group, r.Kind)] = r
+	}
+
+	// Start the CRD listener. When the listener is fully-synced, the listening of actual resources will start.
+	scope.Source.Infof("Beginning CRD Discovery, to figure out resources that are available...")
+	s.provider = rt.NewProvider(s.options.Client, s.options.ResyncPeriod)
+	a := s.provider.GetAdapter(crdKubeResource)
+	s.crdWatcher = newWatcher(crdKubeResource, a)
+	s.crdWatcher.dispatch(event.HandlerFromFn(s.onCrdEvent))
+	s.crdWatcher.start()
+}
+
+func (s *Source) onCrdEvent(e event.Event) {
+	scope.Source.Debuga("onCrdEvent: ", e)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.started {
+		// Avoid any potential timings with .Stop() being called while an event being received.
+		return
+	}
+
+	if s.publishing {
+		// Any event in publishing state causes a reset
+		scope.Source.Infof("Detected a CRD change while processing configuration events. Sending Reset event.")
+		s.handlers.Handle(event.Event{Kind: event.Reset})
+		return
+	}
+
+	switch e.Kind {
+	case event.Added:
+		crd := e.Entry.Item.(*v1beta1.CustomResourceDefinitionSpec)
+		g := crd.Group
+		k := crd.Names.Kind
+		key := asKey(g, k)
+		r, ok := s.expectedResources[key]
+		if ok {
+			scope.Source.Debugf("Marking resource as available: %v", r.CanonicalResourceName())
+			r.Disabled = false
+			s.expectedResources[key] = r
+		}
+
+	case event.FullSync:
```

Comment:
nit: I assume this is the first event that you expect? If so, maybe move it up to the top so that the switch reads in expected order?

Reply:
No, this would be last event to be received under ideal conditions. The switch statement is already in the order of expectation.